File: pkg/model/service_access.go
package model

import "time"

type ServiceAccess struct {
	ID               string     `json:"id" gorm:"primaryKey;size:63"`
	UserID           uint       `json:"-" gorm:"uniqueIndex:idx_service_access_target;not null"`
	Cluster          string     `json:"cluster" gorm:"size:255;uniqueIndex:idx_service_access_target;not null"`
	Namespace        string     `json:"namespace" gorm:"size:63;uniqueIndex:idx_service_access_target;not null"`
	Kind             string     `json:"kind" gorm:"size:16;uniqueIndex:idx_service_access_target;not null"`
	Name             string     `json:"name" gorm:"size:253;uniqueIndex:idx_service_access_target;not null"`
	Port             int        `json:"port" gorm:"uniqueIndex:idx_service_access_target;not null"`
	Scheme           string     `json:"scheme" gorm:"size:5;not null"`
	Path             string     `json:"path" gorm:"type:text;not null"`
	ExpiresInMinutes int        `json:"expiresInMinutes" gorm:"not null;default:180"`
	Public           bool       `json:"public" gorm:"not null;default:false"`
	PublicUntil      *time.Time `json:"publicUntil"`
}

File: ui/src/lib/api/service-access.ts
import { apiClient } from '@/lib/api-client'
import { withSubPath } from '@/lib/subpath'

export interface ServiceAccessEntry {
  id: string
  cluster: string
  namespace: string
  kind: 'services' | 'pods'
  name: string
  port: number
  scheme: string
  path: string
  hostname: string
  expiresAt: string | null
  expiresInMinutes: number
  public: boolean
  publicUntil: string | null
  authorized: boolean
}

export const serviceAccessQueryKey = ['service-access'] as const

export function listServiceAccess() {
  return apiClient.get<ServiceAccessEntry[]>('/service-access')
}

export function deleteServiceAccess(id: string) {
  return apiClient.delete(`/service-access/${encodeURIComponent(id)}`)
}

export function updateServiceAccess(
  id: string,
  settings: { expiresInMinutes: number; public: boolean }
) {
  return apiClient.put(`/service-access/${encodeURIComponent(id)}`, settings)
}

export function serviceAccessOpenURL(id: string) {
  return withSubPath(`/api/v1/service-access/${encodeURIComponent(id)}/open`)
}
